Add cce worldtube reduction

moxcodes opened the pull request:
## Proposed changes

This adds all of the remaining utilities beyond the existing boundary data PRs necessary to perform the conversion of SpEC CCE data to a leaner format that will save a great deal of disk space. On a test file, the data size was reduced by a factor of 3.5 with no compression applied to the reduced data version.

There's a lot of code duplication in this PR, which is intentional to a certain extent. The anticipation is that after we have converted everything SpECTRE needs to interact with to the new format, the legacy data manipulation tools in `ReadBoundaryDataH5.hpp`, the conversion executable, and the adapters for dealing with the SpEC normalization bug (the entire files `SpecBoundaryData.*`, and corresponding test cases from `Test_ReadBoundaryDataH5.cpp`) can be easily snipped from the code base without disrupting any dependencies.

### Depends on: 
- [x] PR #1833
- [x] PR #1809 

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Add reading for reduced format

moxcodes committed:
Add in capability to fix SpEC normalizations

moxcodes committed:
Add conversion script for leaner worldtube data

kidder commented on src/Evolution/Systems/Cce/ReducedWorldtubeModeRecorder.hpp:
```diff
@@ -3,9 +3,13 @@
+
 #include "DataStructures/ComplexModalVector.hpp"
+#include "Evolution/Systems/Cce/Tags.hpp"
 #include "IO/H5/Dat.hpp"
```
remove this include

kidder commented on src/Evolution/Systems/Cce/ReducedWorldtubeModeRecorder.hpp:
```diff
@@ -3,9 +3,13 @@
 #pragma once
 
+#include <cstddef>
+
```
include string, cmath

kidder commented on src/Evolution/Systems/Cce/ReducedWorldtubeModeRecorder.hpp:
```diff
@@ -3,9 +3,13 @@
 
+#include <cstddef>
+
 #include "DataStructures/ComplexModalVector.hpp"
```
just need a forward decl here

kidder reviewed (changes requested):
we are going to wait until you rebase after #1833 is merged as we are confused by what is new...

moxcodes committed:
Add nullptr set_data_ref for `SpinWeighted`

needed for putting `SpinWeighted` in a `Variables`
that is a member variable of a default-constructable class.

moxcodes commented:
This is now squashed, and I added a fixup for the couple of includes comments.
Looks like I've got a file check telling me to get rid of iostream, but I want the progress meter on the conversion process (and it's a non-charm executable). Is there something else I should use or a way to whitelist this executable?
Cheers!

moxcodes commented:
adding back in progress until it compiles properly

moxcodes commented:
Compiler warnings tamed.
Cheers!

kidder commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -113,6 +113,11 @@ struct SpinWeighted<T, Spin, true> {
     data_.set_data_ref(rhs);
   }
 
+  // for invoking the 'check' in `Variables.hpp`
```
please elaborate in this comment what the check is for...

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -221,6 +232,125 @@ class SpecWorldtubeH5BufferUpdater : public WorldtubeBufferUpdater {
+      gsl::not_null<size_t*> time_span_end, double time,
+      size_t interpolator_length, size_t buffer_depth) const noexcept = 0;
+
+  virtual std::unique_ptr<ReducedWorldtubeBufferUpdater> clone_unique() const
```
`clone_unique` -> `get_clone`

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -221,6 +232,125 @@ class SpecWorldtubeH5BufferUpdater : public WorldtubeBufferUpdater {
+    : public ReducedWorldtubeBufferUpdater {
+ public:
+  // charm needs the empty constructor
+  ReducedSpecWorldtubeH5BufferUpdater() noexcept
```
`= default` and default values for the member variables below

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -221,6 +232,125 @@ class SpecWorldtubeH5BufferUpdater : public WorldtubeBufferUpdater {
+
+  explicit ReducedSpecWorldtubeH5BufferUpdater(
+      CkMigrateMessage* /*unused*/) noexcept
+      : extraction_radius_{1.0}, l_max_{0} {}
```
remove the initializers after providing default values

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -221,6 +232,125 @@ class SpecWorldtubeH5BufferUpdater : public WorldtubeBufferUpdater {
+
+  std::unique_ptr<ReducedWorldtubeBufferUpdater> clone_unique() const
+      noexcept override {
+    return std::make_unique<ReducedSpecWorldtubeH5BufferUpdater>(
```
just need the arguments not an object when calling `make_unique`

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -340,7 +470,7 @@ class WorldtubeDataManager {
          Spectral::Swsh::cached_coefficients_metadata(l_max_)) {
       for (size_t i = 0; i < 3; ++i) {
-        for (size_t j = 0; j < 3; ++j) {
+        for (size_t j = i; j < 3; ++j) {
```
this looks like a bug fix, please put it in a separate commit

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -478,4 +608,256 @@ class WorldtubeDataManager {
+          const auto interp_val = interpolator_->interpolate(
+              gsl::span<const double>(time_points.data(), time_points.size()),
+              gsl::span<const std::complex<double>>(
+                  data + column * (buffer_span_size) +
```
why is `buffer_span_size` in `()` ?

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.cpp:
```diff
@@ -107,8 +107,11 @@ double SpecWorldtubeH5BufferUpdater::update_buffers_for_time(
     const gsl::not_null<size_t*> time_span_end, const double time,
     const size_t interpolator_length, const size_t buffer_depth) const
     noexcept {
+  if (*time_span_end >= time_buffer_.size()) {
```
the three changes to this function look like bug fixes; if so, please factor into separate commit

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.cpp:
```diff
@@ -226,7 +230,127 @@ void SpecWorldtubeH5BufferUpdater::update_buffer(
+  const size_t dot_pos = cce_data_filename.find('.');
+  const std::string text_radius =
+      cce_data_filename.substr(r_pos + 1, dot_pos - r_pos - 1);
+  extraction_radius_ = stod(text_radius);
```
how does this error if it fails?

moxcodes replied:
the `stod` should throw an `invalid_argument` exception, which should cause the program to terminate because the function is `noexcept` would you like a catch here with a more detailed `ERROR`?

nilsdeppe replied:
Yes, that would be a lot easier for users.

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -221,6 +232,125 @@ class SpecWorldtubeH5BufferUpdater : public WorldtubeBufferUpdater {
+  DataVector& get_time_buffer() noexcept override { return time_buffer_; }
+
+  /// Serialization for Charm++.
+  void pup(PUP::er& p) noexcept override {
```
move this function to the cpp

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -478,4 +608,256 @@ class WorldtubeDataManager {
+  // charm needs an empty constructor.
+  ReducedWorldtubeDataManager() = default;
+
+  ReducedWorldtubeDataManager(
```
move the constructor definition to the cpp file

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -478,4 +608,256 @@ class WorldtubeDataManager {
+  }
+
+  /// Serialization for Charm++.
+  void pup(PUP::er& p) noexcept {  // NOLINT
```
move the function definition to the cpp

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -478,4 +608,256 @@ class WorldtubeDataManager {
+   * Returns `true` if the time can be supplied from the `buffer_updater_`, and
+   * `false` otherwise. No tags are updated if `false` is returned.
+   */
+  template <typename TagList>
```
move the function definition outside of the class definition

kidder commented on src/Evolution/Systems/Cce/ReducedWorldtubeModeRecorder.hpp:
```diff
@@ -0,0 +1,163 @@
+  /// `modes` rearranged in ascending m-varies-fastest format.
+  ///
+  /// For real quantities, negative m and the imaginary part of m=0 are omitted.
+  void append_worldtube_mode_data(const std::string& dataset_path,
```
put the function definitions into a cpp file

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_ReadBoundaryDataH5.cpp:
```diff
@@ -538,12 +689,178 @@ void test_spec_worldtube_buffer_updater(
+      {value_dist(*gen), value_dist(*gen), value_dist(*gen)}};
+  gr::Solutions::KerrSchild solution{mass, spin, center};
+
+  const double extraction_radius = 100;
```
100.0

kidder reviewed (changes requested):
review of first two commits and fixup

moxcodes committed:
Improve boundary computation numerics

moxcodes committed:
fix off-by-one in `update_buffers_for_time`

moxcodes committed:
use symmetry when reading boundary data components

moxcodes commented:
fixups posted. New non-fixup commits are `Improve boundary computation numerics`, `use symmetry when reading boundary data components`, and `fix off-by-one in update_buffers_for_time`
Cheers!

kidder commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -113,7 +113,8 @@ struct SpinWeighted<T, Spin, true> {
-  // for invoking the 'check' in `Variables.hpp`
+  // needed for invoking the check in `Variables.hpp` that ensures that
+  // default-constructed `Variables` are never used.
   void set_data_ref(std::nullptr_t null, size_t size) noexcept {
```
these arguments should be const

kidder commented on src/Evolution/Systems/Cce/CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ set(LIBRARY_SOURCES
   LinearSolve.cpp
   PrecomputeCceDependencies.cpp
   ReadBoundaryDataH5.cpp
+  ReducedWorldtubeModeRecorder.cpp
```
it appears that the changes to `ReducedWorldtubeMordeRecorder` should be in the fixup commit `fixup: add reading for reduced format` instead of this commit

nilsdeppe replied:
We'll check this once you've squashed since it'll require some git surgery.

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.cpp:
```diff
@@ -80,6 +80,9 @@ SpecWorldtubeH5BufferUpdater::SpecWorldtubeH5BufferUpdater(
   get<Tags::detail::InputDataSet<::Tags::dt<Tags::detail::Lapse>>>(
       dataset_names_) = "/DtLapse";
 
+  radial_derivatives_need_renormalization_ =
```
add comment that `VersionHist` is a SpEC feature and that this is checking whether or not renormalization is required based on whether the data came from an old version of SpEC or not

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -210,6 +214,10 @@ class SpecWorldtubeH5BufferUpdater : public WorldtubeBufferUpdater {
 
+  bool radial_derivatives_need_renormalization() const noexcept override {
+    return radial_derivatives_need_renormalization_;
+  };
```
no semicolon

kidder commented on src/Evolution/Systems/Cce/ReadBoundaryDataH5.hpp:
```diff
@@ -466,6 +475,10 @@ class WorldtubeDataManager {
     // the swsh interface requirement that the spin-weight be labelled with
     // `SpinWeighted`
+
+    // need to separate out the lower spatial metric extraction so that we can
```
we don't understand this comment in relation to the code

kidder commented on src/Evolution/Systems/Cce/SpecBoundaryData.cpp:
```diff
@@ -0,0 +1,275 @@
+
+#include "Evolution/Systems/Cce/BoundaryData.hpp"
+
+#include <cstddef>
```
need Utilities/Gsl.hpp   Tensor/EagerMath/DetAndInverse .  Utilities/ConstantExpressions <complex> Utilities/TMPL

kidder commented on src/Evolution/Systems/Cce/SpecBoundaryData.cpp:
```diff
@@ -0,0 +1,275 @@
+#include "DataStructures/ComplexDataVector.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/SpinWeighted.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
```
not used

kidder commented on src/Evolution/Systems/Cce/SpecBoundaryData.hpp:
```diff
@@ -0,0 +1,301 @@
+
+namespace Cce {
+/*
+ * \brief Compute \f$g_{i j}\f$, \f$g^{i j}\f$, \f$\partial_i g_{j k}\f$, and
```
should these be g or `gamma` as the function name is `spatial_metric`

kidder commented on src/Evolution/Systems/Cce/SpecBoundaryData.hpp:
```diff
@@ -0,0 +1,301 @@
+    size_t l_max) noexcept;
+
+
+template <typename DataBoxTagList>
```
document this function

kidder commented on src/Evolution/Systems/Cce/SpecBoundaryData.hpp:
```diff
@@ -0,0 +1,301 @@
+    const Scalar<ComplexModalVector>& dt_lapse_coefficients,
+    const Scalar<ComplexModalVector>& dr_lapse_coefficients,
+    const double extraction_radius, const size_t l_max) noexcept {
+  size_t size = Spectral::Swsh::number_of_swsh_collocation_points(l_max);
```
const

kidder commented on src/Evolution/Systems/Cce/SpecBoundaryData.hpp:
```diff
@@ -0,0 +1,301 @@
+
+#include <cstddef>
+
+#include "DataStructures/DataBox/DataBox.hpp"
```
include or forward declare for gsl::not_null,   include ComplexDataVector, Variables, Evolution/Systems/Cce/BoundaryData

kidder commented on src/IO/H5/Version.cpp:
```diff
@@ -19,5 +19,18 @@ Version::Version(bool exists, detail::OpenGroup&& group, hid_t location,
       }()),
       group_(std::move(group)) {}
+
+Version::Version(bool exists, detail::OpenGroup&& group, hid_t location,
```
what is this supposed to be doing and what is its intended use case?

moxcodes replied:
This is an overload for writing the version as a string instead of an unsigned int, which is what SpEC does for this type of file.

nilsdeppe replied:
Okay. I guess this can be removed after the catalog is converted? Could you add an explanation to the hpp to the effect of:
```
This is used for fixing a bug in CCE data written by SpEC and will be removed once the CCE data is updated. Do not use for SpECTRE versioning!
```

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_ReadBoundaryDataH5.cpp:
```diff
@@ -45,7 +45,8 @@ void create_fake_time_varying_modal_data(
     const double amplitude, const double frequency, const double time,
-    const size_t l_max, const bool convert_to_goldberg = true) noexcept {
+    const size_t l_max, const bool convert_to_goldberg = true,
+    const bool apply_normalization_bug = false) noexcept {
```
does this apply the bug or the bug_fix?

moxcodes replied:
it applies the bug; it deliberately normalizes incorrectly so that we can check that the new utilities appropriately undo the incorrect normalization.

kidder commented on src/Executables/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 
 add_subdirectory(Benchmark)
 add_subdirectory(DebugPreprocessor)
+add_subdirectory(ReduceCceWorldtube)
```
alphabetize

kidder commented on src/Executables/ReduceCceWorldtube/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(EXECUTABLE reduce_cce_worldtube)
```
CamelCase

kidder commented on src/Executables/ReduceCceWorldtube/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+  Informer
+  GeneralRelativity
+  ${SPECTRE_LIBRARIES}
+  ${YAMLCPP_LIBRARIES}
```
doesn't SPECRE_LIBRARIES already include YAMLCPP_LIBRARIES?

kidder commented on src/Executables/ReduceCceWorldtube/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+target_link_libraries(
+  ${EXECUTABLE}
+  Cce
+  Spectral
```
alphabetize

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+// See LICENSE.txt for details.
+
+#include <boost/program_options.hpp>
+#include <iostream>
```
use Parallel::printf instead of cout

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+#include <iostream>
+#include <string>
+
+#include "DataStructures/DataVector.hpp"
```
is this all the includes you use?

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+
+  for (const auto& libsharp_mode :
+       Spectral::Swsh::cached_coefficients_metadata(computation_l_max)) {
+    if (libsharp_mode.l > l_max) {
```
huh?

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+    if (libsharp_mode.l > l_max) {
+    }
+    for (size_t i = 0; i < 3; ++i) {
+      for (size_t j = 0; j < 3; ++j) {
```
symmetry?

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+
+          Spectral::Swsh::goldberg_modes_to_libsharp_modes_single_pair(
+              libsharp_mode, make_not_null(&spin_weighted_buffer), 0,
+              libsharp_mode.l > l_max
```
factor the ternaries out to variables outside the loops

moxcodes replied:
I don't think I can easily factor them, because the 'else' part of the ternary depends on all of the loop quantities. I'll make it an if statement so that it's more obvious what I'm doing.

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+
+  for (size_t i = 0; i < time_buffer.size(); ++i) {
+    const double time = time_buffer[i];
+    std::cout << "reduceing data at time : " << time << " / "
```
reducing

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+        time_span_end - time_span_start, i - time_span_start, l_max,
+        computation_l_max);
+
+    Cce::create_bondi_boundary_data(
```
where is this function defined; there are too many functions with this name 😢

moxcodes replied:
This one is in `BoundaryData.hpp`, but the question makes me realize I've mistakenly omitted the if for dispatching to the unnormalized version.

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+                                           &output_goldberg_mode_buffer,
+                                           &output_libsharp_mode_buffer, &l_max,
+                                           &computation_l_max,
+                                           &time](auto tag_v) {
```
noexcept

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+  boost::program_options::options_description desc("Options");
+  desc.add_options()("help", "show this help message")(
+      "input_file", boost::program_options::value<std::string>()->required(),
+      "input filename")(
```
name of old CCE data file.

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+
+  boost::program_options::options_description desc("Options");
+  desc.add_options()("help", "show this help message")(
+      "input_file", boost::program_options::value<std::string>()->required(),
```
input_file -> old_spec_cce_file

kidder commented on src/Executables/ReduceCceWorldtube/ReduceCceWorldtube.cpp:
```diff
@@ -0,0 +1,296 @@
+          .run(),
+      vars);
+
+  if (vars.count("help") != 0u) {
```
do this the same way as in Parallel/Main.hpp line 186

kidder commented on src/Executables/ReduceCceWorldtube/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+set_target_properties(
+  ${EXECUTABLE}
+  PROPERTIES LINK_FLAGS "-nomain-module -nomain"
+  )
```
add `add_dependencies(test-executables ${EXECUTABLE})` so this is built

kidder reviewed (changes requested):
joint review with @nilsdeppe

moxcodes commented:
fixups posted.
Cheers!

nilsdeppe commented:
@kidder and I just looked through the fixups and we're happy for you to squash, including any remaining minor changes.

moxcodes commented:
Fixups squashed in -- I think I managed to get everything in the correct commits, but it was many fixups..
Cheers!

kidder commented on src/Evolution/Systems/Cce/ReducedWorldtubeModeRecorder.hpp:
```diff
@@ -59,8 +59,8 @@ SPECTRE_ALWAYS_INLINE std::string dataset_label_for_tag<
 template <>
-std::string
-dataset_label_for_tag<Cce::Tags::BoundaryValue<Cce::Tags::BondiH>>() noexcept {
+SPECTRE_ALWAYS_INLINE std::string dataset_label_for_tag<
```
does this belong in this commit?

moxcodes replied:
you're right - I've moved it to the previous commit.
